File: java/DynamicProgramming_java/LIS/Wiggle_Subsequence_376.java
package java.DynamicProgramming_java.LIS;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Wiggle_Subsequence_376 {

    public int wiggleMaxLength(int[] nums) {
        // its LIS but just a turner if nums[i] grater look te min dp ,if smaller look
        // max dp

        int n = nums.length;
        int ans = 0;
        int[] maxDp = new int[n];
        int[] minDp = new int[n];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] < nums[j]) {
                    maxDp[i] = Math.max(maxDp[i], minDp[j] + 1);
                } else if (nums[i] > nums[j]) {
                    minDp[i] = Math.max(minDp[i], maxDp[j] + 1);
                }

                ans = Math.max(ans, Math.max(maxDp[i], minDp[i]));
            }
        }
        return ans + 1;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Wiggle_Subsequence_376 obj = new Wiggle_Subsequence_376();
            int grid[] = { 1, 17, 5, 10, 13, 15, 10, 5, 16, 8 };
            op.println(obj.wiggleMaxLength(grid));

        }
    }

}

File: Code/DynamicProgramming/LIS/Wiggle_Subsequence_376.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    // your code
    int wiggleMaxLength(vector<int> nums)
    {
        // its LIS but just a turner if  nums[i] grater look te min dp ,if smaller look max dp
        int n = nums.size();
        int ans = 1;
        vector<int> maxDp(n, 1), minDp(n, 1);
        for (int i = 0; i < n; ++i)
        {
            for (int j = 0; j < i; ++j)
            {
                if (nums[i] < nums[j])
                {
                    maxDp[i] = max(maxDp[i], minDp[j] + 1);
                }
                else if (nums[i] > nums[j])
                {
                    minDp[i] = max(minDp[i], maxDp[j] + 1);
                }

                ans = max({ans, maxDp[i], minDp[i]});
            }
        }
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.wiggleMaxLength({1, 17, 5, 10, 13, 15, 10, 5, 16, 8});
    // cout << ans;
    return 0;
}